### PR TITLE
Improve token type support

### PR DESCRIPTION
**Pull request by goerch**
For reference: https://github.com/google/sentencepiece/blob/master/src/model_interface.cc#L193-L212

### CODE REVIEW

**Comment by goerch** [comment]
@ggerganov and others : @klosax proposed [these](https://github.com/ggerganov/llama.cpp/pull/2549#issuecomment-1678016945) additional test cases for the `sentencepiece`-based tokenizer. Are these 'canonical' test cases or do you have different preferences?

Furthermore I'm not sure how to derive the token type for the `BPE`-based tokenizers like `Aquila`?

**Comment by klosax** [comment]
The BPE merges should also be added to the gguf model files like it is done in the falcon / gptneox conversion scripts. User code using more advanced bpe tokenizers needs this.

**Comment by klosax** [comment]
> Furthermore I'm not sure how to derive the token type for the BPE-based tokenizers

Dont know if there is any data for that. The special tokens (EOS / BOS) could be set to control and the rest to normal?

The special token mapping kv from the simpler conversion scripts should also be included. I dont think the llama pth input files have these so should be hardcoded.

**Comment by klosax** [comment]
@goerch status of this? merge eta?

File: tests/test-tokenizer-1.cpp [hunk 1]
```diff
@@ -91,8 +87,8 @@ int main(int argc, char **argv) {
                 return 2;
             }
         } else {
-            if ((vocab_type(ctx) == "spm" && i <= 258) ||
-                (vocab_type(ctx) == "bpe" && (i == 0 || i >= 100000))) {
+            // TODO: needs access to token types
+            if (0 <= i && i < 259) {
```
Comment (ggerganov) [inline tests/test-tokenizer-1.cpp:91]: We can expose token types through the C-style API

e.g.:

```c++
LLAMA_API bool llama_token_is_normal(llama_token id);
...

Would that be enough?
```

**Review by ggerganov — APPROVED** [review]

File: tests/test-tokenizer-1.cpp (hunk 1 above)
Comment (goerch) [reply tests/test-tokenizer-1.cpp:91]: Yes, that would be fine for me.

**Pull request merged**